File: Python Project Filter-Detect-GUI/calculation.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jun  7 11:46:37 2017

@author: Mostafa Hammoud

This function calculates the length of the relevant segments of the ECG

Inputs: 
index : Contains the indices of the PQRST waves
found : Determines whether a wave was detected or misssed
fs    : Sampling frequency
all_ecg    : Contains the Lead I II III filtered
interval   : Conatins RR PR QT QTc Intervals
segment    : Contains PR ST Segments and QRS Complex
value1     : Contains AverageST STIndex IsoIndex Iso STMeasurement of Lead I
value2     : Contains AverageST STIndex IsoIndex Iso STMeasurement of Lead II
value3     : Contains AverageST STIndex IsoIndex Iso STMeasurement of Lead III
duration   : Contains P and T waves durations
three_peaks: Contains the peaks detected in the 3 seconds
ecg_peak   : Contains the Lead I ecg that was used to find the 3 second peaks
start      : The beginnig of the segment
end        : The end of the segment
ecg_peak_lead2: Contains 3 second Lead II
ecg_peak_lead3: Contains 3 second Lead III

Outputs  :
interval : Conatins RR PR QT QTc Intervals 
segment  : Contains PR ST Segments and QRS Complex
value    : Contains AverageST STIndex IsoIndex Iso STMeasurement
duration : Contains P and T waves durations

Additional Comments:
If the value of any of these segments or intervals is [] then it
indicates a missed detection
RR Interval length  is one less than other elements length
"""

import numpy as np
import various_functions

def do_calculation(all_filt_ecg, index, found, fs, interval, segment, value1, 
                   value2, value3, duration, three_peaks, ecg_peak,
                   ecg_peak_lead2, ecg_peak_lead3, start, end):
    
    # Calculate the distance between two consecutive R waves
    # We recalculate for the whole signal since it requires almost no 
    # computation time
    if len(index.r) > 1:
        interval.rr = np.diff(index.r)
    else:
        interval.rr = []
    
    # Set the start index
    begin = len(duration.p)
 
    # Loop through the peaks and do the necessary calculations 
    for i in range(begin,len(index.r)):
        
        # Duration of P wave: poff - pon
        if np.size(index.p[i]) != 0:
            duration.add_p(index.poff[i] - index.pon[i])
        else:
            duration.add_p([])
            
        # Duration of T wave: ton - toff    
        if np.size(index.t[i]) != 0:
            duration.add_t(index.toff[i] - index.ton[i])
        else:
            duration.add_t([])
        
        # PR Interval: qon - pon 
        if np.size(index.p[i]) != 0 and np.size(index.q[i]) != 0:
            interval.add_pr(index.qon[i] - index.pon[i])
        else:
            interval.add_pr([])
        
        # QT Interval: qon - toff
        if np.size(index.q[i]) != 0 and np.size(index.t[i]) != 0:
            interval.add_qt(index.toff[i] - index.qon[i])
        else:
            interval.add_qt([])   
        
        # PR Segment: qon - poff
        if np.size(index.p[i]) != 0 and np.size(index.q[i]) != 0:
            segment.add_pr(index.qon[i] - index.poff[i])
        else:
            segment.add_pr([])
        
        # ST Segment: ton - soff
        if np.size(index.s[i]) != 0 and np.size(index.t[i]) != 0:
            segment.add_st(index.ton[i] - index.soff[i])
            
            st_range = range(index.soff[i],index.ton[i] + 1)
            
            if len(st_range) < 2:                
                value1.add_average_st(np.mean(all_filt_ecg[0][index.ton[i]]))
                value2.add_average_st(np.mean(all_filt_ecg[1][index.ton[i]]))
                value3.add_average_st(np.mean(all_filt_ecg[2][index.ton[i]]))
            else:
                # If Soff is close to S that means it isn't accurate so we will
                # try to fix it
                if index.soff[i] - index.s[i] < 6:
                    temp_start = int(index.soff[i]+index.ton[i])/2
                    st_range = range(int(temp_start),index.ton[i] + 1)
                    
                value1.add_average_st(np.mean(all_filt_ecg[0][st_range]))
                value2.add_average_st(np.mean(all_filt_ecg[1][st_range]))
                value3.add_average_st(np.mean(all_filt_ecg[2][st_range]))
        else:
            
            segment.add_st([])
            value1.add_average_st([])
            value2.add_average_st([])
            value3.add_average_st([])
        
        # QRS Complex(Segment): soff - qon 
        if np.size(index.q[i]) != 0 and np.size(index.s[i]) != 0:
            
            temp = index.soff[i] - index.qon[i]
            
            # In some cases soff and qon are not so accurate so we will use q
            # and s instead add a constant
            if temp > (120*fs/1000):
                temp = index.s[i] - index.q[i] + round(30*fs/1000)
                
            segment.add_qrs(temp)
            
        else:
            segment.add_qrs([])
         
        # Calculate the Iso interval. It is between the end of the current T 
        # wave and the P wave after it
        # Since for the last T wave we need a P wave that has not been found 
        # yet we will use the R wave and try and estimate the range
        
        # The case where we have both T and P
        if i < (len(index.r) - 1):
            
            if np.size(index.t[i]) != 0 and np.size(index.p[i+1]) != 0:
                
                iso_range = range(index.toff[i],index.pon[i+1] + 1)
                
                iso = np.mean(all_filt_ecg[0][iso_range])
                value1.add_iso(iso)
                iso = np.mean(all_filt_ecg[1][iso_range])
                value2.add_iso(iso)
                iso = np.mean(all_filt_ecg[2][iso_range])
                value3.add_iso(iso)
                
            else:
                value1.add_iso([])
                value2.add_iso([])
                value3.add_iso([])
          
        # The case where we dont have the P wave   
        else:
            
            # Use the peaks that were detected in the 3 second range
           last_peaks =  various_functions.remove_unwanted(three_peaks,
                                                           end, end + fs)
           
           if last_peaks != []:
               # calculate the RR range to get p offset widows
               rr = np.diff(three_peaks[:-2])
               
               if np.size(rr) > 0:
                   rr = np.mean(rr)
                   
                   p_window_start = int(np.round(rr*0.25))
                   pon_window     = int(np.round(rr*0.08))
                   
               else:
                   p_window_start = int(np.round(200*fs/1000))
                   pon_window     = int(np.round(60 *fs/1000))                   
                
               # Specify the Iso range which starts at the end of the T wave 
               # and ends at (R_index - P_window - Pon_window)
               iso_range = np.arange(index.toff[i] - start,last_peaks[0] - 
                                 p_window_start - pon_window - start)
               
               # If the Iso range is small or too big its not relaible
               if np.size(iso_range) > 2 and np.size(iso_range) < 80:
                   # Calculte the mean of the Iso range
                  value1.add_iso(np.mean(ecg_peak[iso_range]))
                  value2.add_iso(np.mean(ecg_peak_lead2[iso_range]))
                  value3.add_iso(np.mean(ecg_peak_lead3[iso_range]))
               else:
                   value1.add_iso([])
                   value2.add_iso([])
                   value3.add_iso([])
           else:
               value1.add_iso([])
               value2.add_iso([])
               value3.add_iso([])

        if np.size(value1.iso[i]) != 0 and np.size(segment.st[i]) != 0:
            value1.add_st_measurement(value1.average_st[i] - value1.iso[i])
            value2.add_st_measurement(value2.average_st[i] - value2.iso[i])
            value3.add_st_measurement(value3.average_st[i] - value3.iso[i])
        else:
            value1.add_st_measurement([])
            value2.add_st_measurement([])
            value3.add_st_measurement([])
            
    # Select the start for the QTc interval  
    begin = len(interval.qtc)
    
    # Loop through the QT interval and perfom Bazett's calculation
    # We need to use the RR interval for that
    for i in range(begin,len(interval.qt)):
        if len(interval.rr) >= 1:
            # Since RR interval is smaller than the QT interval
            # For the last QTc calculation we will use the last RR interval 
            # So the last RR interval will be used twice
            if i < len(interval.rr):
                interval.add_qtc(interval.qt[i]/np.sqrt(interval.rr[i]))
            else:
                interval.add_qtc(interval.qt[i]/np.sqrt(interval.rr[-1]))
                
        else:
            interval.add_qtc([])

          
            

File: Matlab/Calculation.m
function [ Interval, Segment, Value, Duration, Count ] = Calculation....
    ( Index,Fail,fs,Buffer,Interval,Segment,Value,Duration,Count,IsAstro)
% This function calculates the length of the relevant segments of the ECG
%
% Inputs: 
% Index : Contains the indices of the PQRST waves
% Fail  : Determines whether a wave was detected or misssed
% fs    : Sampling frequency
%
% Outputs  :
% Interval : Conatins RR PR QT QTc Intervals
% Segment  : Contains PR ST Segments and QRS Complex
% Value    : Contains AverageST STIndex IsoIndex Iso STMeasurement
% Duration : Contains P and T waves durations
%
% Additional Comments:
% If the value of any of these segments or intervals is zero then it
% indicates a missed detection
% RR Interval Iso and STMeasurement length  is one less than other elements
% length


%% Calculate RR Interval
% It's the difference between two consecutive peaks
Interval.RR = diff(Index.R);
Interval.RRIndex = Index.R(1:end-1);

% Counters to keep track of the waves

% Used in P wave Duration
PDurCount = Count.P;
% Used in T wave Duration
TDurCount = Count.T;
% Used in PR Interval
PRCountP  = Count.PRP;
PRCountQ  = Count.PRQ;
PRCount   = Count.PR;
% Used in QT Interval
QTCountQ  = Count.QTQ;
QTCountT  = Count.QTT;
QTCount   = Count.QT;
% Used in PR Segment
PRSCountP = Count.PRSP;
PRSCountQ = Count.PRSQ;
PRSCount  = Count.PRS;
% Used in ST Segment
STSCountS = Count.STSS;
STSCountT = Count.STST;
STSCount  = Count.STS;
% Used in QRS Complex
QRSCountQ = Count.QRSQ;
QRSCountS = Count.QRSS;
QRSCount  = Count.QRS;
% Used in Isoelectric Baseline
IsoCountT = Count.IsoT;
IsoCountP = Count.IsoP;
IsoCount  = Count.Iso;

% Beginning of the loop
Start = length(Duration.P) + 1;

% When this loop runs more than once the last element of the old loop will
% overlap with the first element of the new loop. This is intentional since
% in the case of Iso calculation we need two consecutive heart beats so we
% will use the last old heart beat and the first new heart beat
for i = Start:length(Index.R)
    
    %% Calculate P wave duration
    if Fail.P(i) == 1
        PDurCount = PDurCount + 1;
        
        % Get number of missed P waves before i location
        
        
        Duration.P(i) = Index.Poff(PDurCount) - Index.Pon(PDurCount);
    else
        Duration.P(i) = 0;
    end
    
    %% Calculate T wave duration
    if Fail.T(i) == 1
        TDurCount = TDurCount + 1;
        Duration.T(i) = Index.Toff(TDurCount) - Index.Ton(TDurCount);
    else
        Duration.T(i) = 0;
    end
    
    %% Calculate PR Interval
    % It begins at the onset of a P wave -> onset of a Q wave
    
    % If both Q and T waves were found do the calculation
    if Fail.P(i) + Fail.Q(i) == 2
        
        % Counter for everytime the if statement was valid
        PRCount = PRCount + 1;
        
        % Store the difference in i location
        Interval.PR(i) = Index.Qon(PRCount + PRCountQ) - ...
            Index.Pon(PRCount + PRCountP);
        
        Interval.PRIndex(i) = Index.Pon(PRCount + PRCountP);
    else
        % Cases where P or Q was not found will be filled with zero
        Interval.PR(i)      = 0;
        Interval.PRIndex(i) = 0;
        
        % If a P wave was found but a Q wave was missed
        if Fail.P(i) == 1
           PRCountP = PRCountP + 1;
        % If a Q wave was found but a P wave was missed
        elseif Fail.Q(i) == 1
            PRCountQ = PRCountQ + 1;
        end
    end
    
    %% Calculate QT Interval
    % It begins at the onset of a Q wave-> endpoint of a T wave
    
    % If both Q and T waves were found do the calculation
    if Fail.Q(i) + Fail.T(i) == 2
        
        % Counter for everytime the if statement was valid
        QTCount = QTCount + 1;
        
        % Store the difference in i location
        Interval.QT(i) = Index.Toff(QTCount + QTCountT) - ...
            Index.Qon(QTCount + QTCountQ);
        
        Interval.QTIndex(i) = Index.Qon(QTCount + QTCountQ);

    else
        % Cases where Q or T were not found will be filled with zero
        Interval.QT(i)      = 0;
        Interval.QTIndex(i) = 0;
        
        % If a P wave was found but a Q wave was missed
        if Fail.T(i) == 1
           QTCountT = QTCountT + 1;
        % If a Q wave was found but a P wave was missed
        elseif Fail.Q(i) == 1
            QTCountQ = QTCountQ + 1;
        end
        
    end
    
    %% Calculate PR Segment
    % It begins at the endpoint of a P wave -> onset of a Q wave
    
    % If both P and Q waves were found do the calculation
    if Fail.P(i) + Fail.Q(i) == 2
        
        % Counter for everytime the if statement was valid
        PRSCount = PRSCount + 1;        
        
        % Store the difference in i location
        Segment.PR(i) = Index.Qon(PRSCount + PRSCountQ) - ...
            Index.Poff(PRSCount + PRSCountP);
        Segment.PRIndex(i) = Index.Poff(PRSCount + PRSCountP);
    else
        % Cases where P or Q were not found will be filled with zero
        Segment.PR(i) = 0;
        Segment.PRIndex(i) = 0;
        
        % If a P wave was found but a Q wave was missed
        if Fail.P(i) == 1
           PRSCountP = PRSCountP + 1;
        % If a Q wave was found but a P wave was missed
        elseif Fail.Q(i) == 1
            PRSCountQ = PRSCountQ + 1;
        end
        
    end
    
    %% Calculate ST Segment and Average ST
    % It begins at the endpoint of an S wave -> onset of a T wave
    
    % If both S and T waves were found do the calculation
    if Fail.S(i) + Fail.T(i) == 2
        
        % Counter for everytime the if statement was valid
        STSCount = STSCount + 1;   
        
        % Store the difference in i location
        Segment.ST(i) = Index.Ton(STSCount + STSCountT) - ...
            Index.Soff(STSCount + STSCountS);
        Segment.STIndex(i) = Index.Soff(STSCount + STSCountS);
        
        % Calculate the range for averging
        STRange = Index.Soff(STSCount + STSCountS):...
            Index.Ton(STSCount + STSCountT);

        % If STRange is so low take Ton as the AverageST
        if length(STRange) < 2
            % Find the mean of the ECG during the specific range
            Value.AverageST(i) = mean(Buffer.Plot(Index.Ton(STSCount...
                + STSCountT)));
            Value.STIndex(i)   = Index.Soff(STSCount + STSCountS);
        else
            % Find the mean of the ECG during the specific range
            Value.AverageST(i) = mean(Buffer.Plot(STRange));
            Value.STIndex(i)   = Index.Soff(STSCount + STSCountS);
        end
        
    else
        % Cases where S or T were not found will be filled with zero
        Segment.ST(i)      = 0;
        Segment.STIndex(i) = 0;
        Value.AverageST(i) = 0;
        Value.STIndex(i)   = 0;
        
        % If a P wave was found but a Q wave was missed
        if Fail.T(i) == 1
           STSCountT = STSCountT + 1;
        % If a Q wave was found but a P wave was missed
        elseif Fail.S(i) == 1
            STSCountS = STSCountS + 1;
        end
    end
    
    %% Calculate QRS Complex
    % It begins at the onset of a Q wave -> endpoint of an S wave
    
    % If both Q and S waves were found do the calculation
    if Fail.Q(i) + Fail.S(i) == 2
        
        % Counter for everytime the if statement was valid
        QRSCount = QRSCount + 1;
        
        % Store the difference in i location
        Segment.QRS(i) = Index.Soff(QRSCount + QRSCountS) - ...
            Index.Qon(QRSCount + QRSCountQ);
        Segment.QRSIndex(i) = Index.Qon(QRSCount + QRSCountQ);
        if IsAstro
        if Segment.QRS(i) > (120*fs/1000)
            Segment.QRS(i) = Index.S(QRSCount + QRSCountS) - ...
                Index.Q(QRSCount + QRSCountQ) + round(30*fs/1000);
            Segment.QRSIndex(i) = Index.Q(QRSCount + QRSCountQ);
        end
        end
        
    else
        % Cases where Q or S were not found will be filled with zero
        Segment.QRS(i) = 0;
        Segment.QRSIndex(i) = 0;
        
        % If a P wave was found but a Q wave was missed
        if Fail.Q(i) == 1
           QRSCountQ = QRSCountQ + 1;
        % If a Q wave was found but a P wave was missed
        elseif Fail.S(i) == 1
            QRSCountS = QRSCountS + 1;
        end
        
    end
    
    %% Calculate Isoelectric Baseline
    % It is the flat part of the ECG
    % Calculated by averging the points between two consecutive heart beats
    % Mean of points between endpoint of a T wave -> on set of a P wave
    
    Loc = i;
    Bool = true;
    TempIsoCount  = IsoCount ;
    TempIsoCountT = IsoCountT;
    
    % When the end of the loop arrives find Iso and ST measurement from the
    % difference between the waves of first new loop and last old loop
    if i == length(Index.R)
        Loc = Start - 1;

        
        % If loc < 1 then the detection just started we cannot check with
        % old pulses
        if Loc < 1
            Bool = false;
            Loc = 1;
            
            if Fail.T(i) == 1
                TempIsoCountT = TempIsoCountT + 1;
            end
        
        else
            NewIsoCount   = TempIsoCount;
            if Fail.T(Loc) == 1
                NewIsoCountT  = TempIsoCountT - 1;
            else
                NewIsoCountT  = TempIsoCountT;
            end
            TempIsoCount  = Count.Iso;
            if Fail.T(Loc) == 1
                TempIsoCountT = Count.IsoT - 1;
            else
                TempIsoCountT = Count.IsoT;
            end
        end

    end
    % Prevent negative indexing
    if length(Fail.P) >= Loc+1
    if (Fail.T(Loc) + Fail.P(Loc+1) == 2 ) && Bool
        
        % Counter for everytime the if statement was valid
        TempIsoCount = TempIsoCount + 1;
        
        if exist('NewIsoCount','var')
            NewIsoCount  = NewIsoCount + 1;
        end
        
        Temp = Index.Pon>Index.Toff(TempIsoCount + TempIsoCountT) &...
            Index.Pon<Index.Toff(TempIsoCount + TempIsoCountT) + 500*fs/1000;
        
        IsoRange = Index.Toff(TempIsoCount + TempIsoCountT):...
            Index.Pon(Temp);
        
        Value.Iso(Loc)      = mean(Buffer.Plot(IsoRange));
        Value.IsoIndex(Loc) = Index.Toff(TempIsoCount + TempIsoCountT);
        
    elseif Bool
        % Cases where P or T were not found will be filled with zero
        Value.Iso(Loc)      = 0;
        Value.IsoIndex(Loc) = 0;
        
       
        if exist('NewIsoCountT','var') && Fail.T(Loc) == 1
            NewIsoCountT  = NewIsoCountT + 1;
        end
        
        if Fail.T(Loc) == 1
            TempIsoCountT = TempIsoCountT + 1;
        end     
    end
    end


    if exist('NewIsoCount','var')
        IsoCount  = NewIsoCount;
    else
        IsoCount  = TempIsoCount;
    end
    
    if exist('NewIsoCountT','var')
        if Fail.T(i) == 1 %&& Fail.P(i) == 1
            IsoCountT  = NewIsoCountT + 1;
        else
            IsoCountT  = NewIsoCountT;
        end
            
    else
        IsoCountT  = TempIsoCountT;
    end

    
    %% Calculate ST Measurement(Elevation - Depression)
    % It is AverageST - Isoelectric Baseline
    
    % Prevent negative indexing
    if length(Value.Iso) >= Loc
    if Value.Iso(Loc) ~= 0 && Segment.ST(Loc)~= 0 
        Value.STMeasurement(Loc) = Value.AverageST(Loc) - Value.Iso(Loc);
    else
        Value.STMeasurement(Loc) = 0;
    end
    else
        Value.STMeasurement(Loc) = 0;
    end
end


%% Perform Bazett's calculation for QTc
% Cases were Q or T were not found will be filled with zero

Start = length(Interval.QTc) + 1;

for i = Start:length(Interval.QT)
    if length(Interval.RR) >= 1
    if i < length(Interval.RR)
        Interval.QTc(i) = (Interval.QT(i))/(sqrt(Interval.RR(i)/fs));
    else
        Interval.QTc(i) = (Interval.QT(i))/(sqrt(Interval.RR(end)/fs));
    end
    else
        Interval.QTc(i) = 0;
    end
end

%% Save the counters
Count.P    = PDurCount;
Count.T    = TDurCount;
Count.PRP  = PRCountP ;
Count.PRQ  = PRCountQ ;
Count.PR   = PRCount  ;
Count.QTQ  = QTCountQ ;
Count.QTT  = QTCountT ;
Count.QT   = QTCount  ;
Count.PRSP = PRSCountP;
Count.PRSQ = PRSCountQ;
Count.PRS  = PRSCount ;
Count.STSS = STSCountS;
Count.STST = STSCountT;
Count.STS  = STSCount ;
Count.QRSQ = QRSCountQ;
Count.QRSS = QRSCountS;
Count.QRS  = QRSCount ;
Count.IsoT = IsoCountT;
Count.IsoP = IsoCountP;
Count.Iso  = IsoCount ;

end










